<a href="https://colab.research.google.com/github/Alonge9500/Alonge9500/blob/main/logisticVsSinglePerceptron.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### A demonstration that binomial logistic regression is equivalent to a single layer perceptron with a cross entropy loss function Using the titanic Dataset

### Steps:
* Load Data
* Drop Na
* Clean Data
* Standardize Features
* Build a Logistic Regression Model Using Scikit Learn
* Build a single layer Perceptron using keras
* Compare the Log Loss
* Conclusion

In [ ]:
import pandas as pd
from sklearn import preprocessing
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, log_loss
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

In [ ]:
titanic = pd.read_csv("https://web.stanford.edu/class/archive/cs/cs109/cs109.1166/stuff/titanic.csv")
titanic = titanic.dropna()
titanic['Sex'] = titanic['Sex'].replace(['female', 'male'], [0, 1])
titanic = titanic.drop(['Name'], axis=1)



In [ ]:
titanic.head()

Survived  Pclass  Sex   Age  Siblings/Spouses Aboard  \
0         0       3    1  22.0                        1   
1         1       1    0  38.0                        1   
2         1       3    0  26.0                        0   
3         1       1    0  35.0                        1   
4         0       3    1  35.0                        0   

   Parents/Children Aboard     Fare  
0                        0   7.2500  
1                        0  71.2833  
2                        0   7.9250  
3                        0  53.1000  
4                        0   8.0500

In [ ]:
#Split and Scale Data

X = titanic.drop(columns=['Survived'])
y = titanic['Survived']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
#Logistic Regression
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)
y_pred_prob = log_reg.predict_proba(X_test)[:, 1]
y_pred = log_reg.predict(X_test)
log_reg_accuracy = accuracy_score(y_test, y_pred)
log_reg_log_loss = log_loss(y_test, y_pred_prob)

In [ ]:
X_train.shape

(709, 6)

In [ ]:
# Single-Layer Perceptron using Cross-Entropy Loss
model = Sequential([Dense(1, input_shape=(6,), activation='sigmoid')])
model.compile(optimizer=Adam(learning_rate=0.01), loss='binary_crossentropy', metrics=['accuracy'])
history = model.fit(X_train, y_train, epochs=100, validation_split=0.15, verbose=0)
loss, perceptron_accuracy = model.evaluate(X_test, y_test, verbose=0)
y_pred_for_perceptron = model.predict(X_test).flatten()
y_pred_perceptron = (y_pred_for_perceptron > 0.5).astype(int)
logloss_for_perceptron = log_loss(y_test, y_pred_for_perceptron)


6/6 [==============================] - 0s 2ms/step


In [ ]:

# Comparison
print(f"Logistic Regression Accuracy: {log_reg_accuracy}")
print(f"Logistic Regression Log Loss: {log_reg_log_loss}")
print(f"Single-Layer Perceptron Accuracy: {perceptron_accuracy}")
print(f"Single-Layer Perceptron Log Loss: {logloss_for_perceptron}")

Logistic Regression Accuracy: 0.7471910112359551
Logistic Regression Log Loss: 0.5045658004292629
Single-Layer Perceptron Accuracy: 0.7471910119056702
Single-Layer Perceptron Log Loss: 0.5057085499706402


### **Conclusion**
* Base on the result above the Log Loss for the Logistic regression is 0.505 (3dp) while the Log Loss for the Single Layer Perceptron is 0.506(3dp) and the accuracy for the logistic regression and single layer perceptron are 0.747 and 0.747 (3dp) Respectively. Hence we can conclude that:
**binomial logistic regression is equivalent to a single layer perceptron with a cross entropy loss function**